src/main/java/com/google/cloud/tools/managedcloudsdk/install/Extractor.java

```diff
@@ -16,42 +16,50 @@
 
 package com.google.cloud.tools.managedcloudsdk.install;
 
-import com.google.cloud.tools.managedcloudsdk.MessageListener;
+import com.google.cloud.tools.managedcloudsdk.ProgressListener;
 import com.google.common.annotations.VisibleForTesting;
 import java.io.IOException;
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.util.logging.Logger;
 
 /**
  * Extractor for extracting files from a single archive. Use {@link ExtractorProvider} to provide
  * extractor implementation.
  */
 final class Extractor<T extends ExtractorProvider> {
+
+  private final Logger logger = Logger.getLogger(Extractor.class.getName());
+
   private final Path archive;
   private final Path destination;
   private final ExtractorProvider extractorProvider;
-  private final MessageListener messageListener;
+  private final ProgressListener progressListener;
 
   /** Use {@link ExtractorFactory} to instantiate. */
-  Extractor(Path archive, Path destination, T extractorProvider, MessageListener messageListener) {
+  Extractor(
+      Path archive, Path destination, T extractorProvider, ProgressListener progressListener) {
     this.archive = archive;
     this.destination = destination;
     this.extractorProvider = extractorProvider;
-    this.messageListener = messageListener;
+    this.progressListener = progressListener;
   }
 
   /** Extract an archive. */
   public void extract() throws IOException, InterruptedException {
-    messageListener.message("Extracting archive: " + archive + "\n");
+    progressListener.start("Extracting archive: " + archive.getFileName(), 2);
+    // The extractor is responsible for progress from [100 to 200], if you plan on reusing this,
+    // pass the offset in as a parameter.
+    progressListener.update(1);
 
     try {
-      extractorProvider.extract(archive, destination, messageListener);
+      extractorProvider.extract(archive, destination);
```

Comment:
It would be nice to provide a child ProgressListener here, but I know the zip and tar.gz extractors don't support obtaining a file count.  The Eclipse `IProgressMonitor`'s support an `UNKNOWN` (-1) as _totalWork_, in which case `update()`s are reported upwards using some kind of exponential backoff (IIRC).  It keeps the progress bar alive.  I incorporated this idea into an alternative child ProgressListener below.

Reply:
I mean -1 could be written into the contract, so implements will presumably have to handle/deal with it if we chose to do it?

Reply:
Yeah: written into the contract.  TarGz and Zip extractor would call `start("extracting files", ProgressListener.UNKNOWN)` and just call `update(1)` after each file.  On a non-SSD (or in a VM), extracting the Cloud SDK is a visible chunk of work.

Reply:
https://github.com/GoogleCloudPlatform/appengine-plugins-core/issues/529